### PR TITLE
Segment layers and better effect transitions (blending)

### PR DESCRIPTION
**Pull request by blazoncek**
Segment layering & effect blending improvements
- Photoshop-style segment/layer blending
- return of strip ABL
- remove global LED buffer in favour of segment-local buffer
- new effect blending modes/transitions
- custom palettes moved out of WS2812FX class
- increased limits (matrix size, LED RAM)
- added "rainbow"-mode colorwheel
- replaced palettes with gamma unmodified ones
- move gamma adjustment to last step before sending to LEDs
- Segment & WS2812FX class reorganisation (mutable members, reordered members, protected members)

Fulfils #4550 
Fixes: #4610 #4603 (may) #4553 #4520 #4140 #3712 #3417 

## Summary by CodeRabbit

- **New Features**
  - Added new 2D segment blend modes including swipe and circular transitions, accessible via updated UI controls.
  - Introduced loading of custom color palettes from JSON files for enhanced customization.
  - Added a "rainbow" color wheel option alongside the default spectrum wheel.
  - Enhanced global palette blending and wrapping settings with multiple modes.

- **Improvements**
  - Optimized memory management for LED data, presets, MQTT, and file operations, leveraging PSRAM/DRAM on supported platforms.
  - Improved 2D matrix setup, panel management, and segment transition handling for smoother effects.
  - Updated default color palettes with richer, more vibrant gradients.
  - Streamlined configuration and preset loading with better defaults and pin validation.
  - Refined segment value application and transition handling for improved brightness and effect consistency.
  - Simplified realtime and overlay pixel handling for better performance and clarity.

- **Bug Fixes**
  - Fixed palette count references in UI, IR remote, and button controls for consistency.
  - Corrected segment and pixel color handling in overlays and effects.
  - Resolved issues with segment transitions, effect resets, and state change detection.
  - Addressed timing debug output formatting for accurate logging.

- **User Interface**
  - Redesigned blend mode selection UI with new options and improved styling.
  - Moved and added settings for palette blending and color wheel preferences in LED settings.
  - Enhanced segment and panel configuration displays for better clarity and usability.
  - Updated holiday list and tooltip formatting in the web UI.

- **Refactor**
  - Improved encapsulation and const correctness in segment and pixel management.
  - Removed deprecated global LED buffer feature and related UI elements.
  - Modernized function signatures and internal logic for better maintainability.
  - Standardized return types and memory allocation wrappers across modules.
  - Reorganized includes and cleaned up legacy macros and type usage.

- **Chores**
  - Updated documentation comments and fixed URLs.
  - Cleaned up legacy macros, type usage, and includes for consistency and clarity.

### CODE REVIEW

**Comment by blazoncek** [comment]
There are a lot of changes in Segment and WS2812FX classes so it may be too tedious to go through diff.

The limitations are mainly based on amount of free RAM and ESP8266 may only work with 2 to 3 segments of 256 pixels (16x16 matrix) overlaid. S2 may struggle too if many usermods are loaded.
I've marked PR draft until @DedeHai combs through particle system and makes sure it plays nicely with new code and all missing spots are found. I suggest code is modified in this PR (layers branch) as every collaborator has access.

The side effect of the new approach is more than 3x speed improvement compared to old versions.

**Comment by DedeHai** [comment]
I started reviewing, I saw a lot of changes unrelated to layering, in scrolling text for example. Would it make sense to keep those to a minimum? I can  also add comments and suggestions.

**Comment by blazoncek** [comment]
> I saw a lot of changes unrelated to layering

Quite true, however, as I said previously, these are collateral damage from developing in my fork. All of them are intended to make WLED code cleaner and introduce "separations of duties". I do not like interdependencies and prefer modularity.

As for Scrolling text, the change does not affect final output. As a matter of fact I have modification to "Overlay" option for other effects ready. It is no longer needed and can be removed as any effect can be overlaid. Some other workarounds implemented (i.e. Matrix effect etc) can be removed too.

**Comment by DedeHai** [comment]
Ok, I keep that in mind during the review and will ad my 2 cents :)

File: wled00/led.cpp [hunk 1]
```diff
@@ -94,7 +87,7 @@ void applyFinalBri() {
 void stateUpdated(byte callMode) {
   //call for notifier -> 0: init 1: direct change 2: button 3: notification 4: nightlight 5: other (No notification)
   //                     6: fx changed 7: hue 8: preset cycle 9: blynk 10: alexa 11: ws send only 12: button preset
-  setValuesFromFirstSelectedSeg();
+  setValuesFromFirstSelectedSeg();  // a much better approach would be to use main segment: setValuesFromMainSeg()
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/led.cpp:91]: Untrue.
`getSegment()` guards against out-of-bounds segment ID. Also `getFirstSelectedSegId()` falls back to `getMainSegmentId()` if no segments are selected.

File: wled00/led.cpp [hunk 2]
```diff
@@ -24,25 +22,19 @@ void setValuesFromSegment(uint8_t s)
...
+    if (effectIntensity != seg.intensity) {seg.intensity = effectIntensity; stateChanged = true;}
+    if (effectPalette   != seg.palette)   {seg.setPalette(effectPalette);}
+    if (effectCurrent   != seg.mode)      {seg.setMode(effectCurrent);}
     uint32_t col0 = RGBW32(colPri[0], colPri[1], colPri[2], colPri[3]);
     uint32_t col1 = RGBW32(colSec[0], colSec[1], colSec[2], colSec[3]);
-    if (col0 != selsegPrev.colors[0])            {seg.setColor(0, col0);}
-    if (col1 != selsegPrev.colors[1])            {seg.setColor(1, col1);}
+    if (col0 != seg.colors[0])            {seg.setColor(0, col0);}
+    if (col1 != seg.colors[1])            {seg.setColor(1, col1);}
   }
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/led.cpp:38]: Untrue. `setColor()` does that.

File: wled00/bus_manager.cpp [hunk 3]
```diff
@@ -832,22 +787,24 @@ unsigned BusManager::memUsage() {
...
+  }
+  if (digital > WLED_MAX_DIGITAL_CHANNELS || analog > WLED_MAX_ANALOG_CHANNELS) return -1;
   if (Bus::isVirtual(bc.type)) {
     busses.push_back(make_unique<BusNetwork>(bc));
-    //busses.push_back(new BusNetwork(bc));
   } else if (Bus::isDigital(bc.type)) {
-    busses.push_back(make_unique<BusDigital>(bc, numDigital));
-    //busses.push_back(new BusDigital(bc, numDigital));
+    busses.push_back(make_unique<BusDigital>(bc, Bus::is2Pin(bc.type) ? twoPin : digital));
   } else if (Bus::isOnOff(bc.type)) {
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/bus_manager.cpp:804]: Only makes sense for analog.

File: wled00/FX_fcn.cpp [hunk 4]
```diff
@@ -1039,36 +882,42 @@ void Segment::setPixelColor(float i, uint32_t col, bool aa) const
 
 uint32_t IRAM_ATTR_YN Segment::getPixelColor(int i) const
 {
-  if (!isActive()) return 0; // not active
+  if (!isActive() || i < 0) return 0; // not active or invalid index
```
Comment (DedeHai) [inline wled00/FX_fcn.cpp:885]: see comment above: if negative is invalid, make it unsigned.

File: wled00/FX_fcn.cpp [hunk 5]
```diff
@@ -1079,157 +928,69 @@ uint32_t IRAM_ATTR_YN Segment::getPixelColor(int i) const
...
-  const int rows = vHeight(); // will be 1 for 1D
-  // pre-scale color for all pixels
-  c = color_fade(c, _segBri);
-  _colorScaled = true;
-  for (int y = 0; y < rows; y++) for (int x = 0; x < cols; x++) {
-    if (is2D()) setPixelColorXY(x, y, c);
-    else        setPixelColor(x, c);
-  }
-  _colorScaled = false;
+  for (unsigned i = 0; i < length(); i++) setPixelColorRaw(i,c); // always fill all pixels (blending will take care of grouping, spacing and clipping)
```
Comment (DedeHai) [inline wled00/FX_fcn.cpp:980]: this is so much nicer now.

File: wled00/FX_fcn.cpp [hunk 6]
```diff
@@ -1286,27 +1034,27 @@ void Segment::blur(uint8_t blur_amount, bool smear) {
   }
 #endif
   uint8_t keep = smear ? 255 : 255 - blur_amount;
-  uint8_t seep = blur_amount >> 1;
+  uint8_t seep = blur_amount >> (1 + smear);
```
Comment (DedeHai) [inline wled00/FX_fcn.cpp:1037]: please dont add this back in.

File: wled00/FX_fcn.cpp [hunk 7]
```diff
@@ -1286,27 +1034,27 @@ void Segment::blur(uint8_t blur_amount, bool smear) {
...
   unsigned vlength = vLength();
   uint32_t carryover = BLACK;
-  uint32_t lastnew;       // not necessary to initialize lastnew and last, as both will be initialized by the first loop iteration
```
Comment (DedeHai) [inline wled00/FX_fcn.cpp:1292]: dont remove this comment. its there deliberately.

File: wled00/FX_fcn.cpp [hunk 8]
```diff
@@ -1315,17 +1063,23 @@ void Segment::blur(uint8_t blur_amount, bool smear) {
...
-    pos -= 85;
-    return RGBW32(0, (pos * 3), (255 - pos * 3), w);
+  if (useRainbowWheel) {
+    CRGB rgb;
+    hsv2rgb_rainbow(CHSV(pos, 255, 255), rgb);
+    return RGBW32(rgb.r, rgb.g, rgb.b, w);
   } else {
-    pos -= 170;
-    return RGBW32((pos * 3), (255 - pos * 3), 0, w);
+    if (pos < 85) {
```
Comment (DedeHai) [inline wled00/FX_fcn.cpp:1074]: is hsv2rgb_spectrum() any different from this?
also:
I am not sure its necessary to differentiate it at all, could alsways use rainbow, it gives nicer colors and is actually the default for hsv->rgb conversions in fastled. I assume you added this for legacy reasons and user complaints? Did you check if the two options give any noticeable differences in the FX that use it? If the difference is really small, this could be a compile parameter for the edge cases where users would want "the old look". what do you think?

**Review by DedeHai — CHANGES_REQUESTED** [review]
partially done reviewing. still need to go over it some more and then run some tests.

File: wled00/FX.cpp [hunk 9]
```diff
@@ -5053,7 +5053,11 @@ uint16_t mode_2Dfirenoise(void) {               // firenoise2d. By Andrew Tuline
   unsigned yscale = SEGMENT.speed*8;
   unsigned indexx = 0;
 
-  CRGBPalette16 pal = SEGMENT.check1 ? SEGPALETTE : SEGMENT.loadPalette(pal, 35);  
+  //CRGBPalette16 pal = SEGMENT.check1 ? SEGPALETTE : SEGMENT.loadPalette(pal, 35);  
+  CRGBPalette16 pal = SEGMENT.check1 ? SEGPALETTE : CRGBPalette16(CRGB::Black,     CRGB::Black,      CRGB::Black,  CRGB::Black,
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX.cpp:5057]: `loadPalette()` is inaccessible now (intentionally). Either move palette into pre-defined palettes or create a palette on the fly (as it was originally).

File: wled00/FX.cpp [hunk 10]
```diff
@@ -6159,27 +6165,28 @@ uint16_t mode_2Dscrollingtext(void) {
     SEGENV.step = strip.now + map(SEGMENT.speed, 0, 255, 250, 50); // shift letters every ~250ms to ~50ms
   }
 
-  if (!SEGMENT.check2) SEGMENT.fade_out(255 - (SEGMENT.custom1>>4));  // trail
-  bool usePaletteGradient = false;
+  SEGMENT.fade_out(255 - (SEGMENT.custom1>>4));  // trail
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX.cpp:6168]: Check2 was "Overlay" which is no longer used.

File: wled00/FX_fcn.cpp [hunk 11]
```diff
@@ -10,7 +10,6 @@
   Modified heavily for WLED
 */
 #include "wled.h"
-#include "FX.h"
 #include "FXparticleSystem.h"  // TODO: better define the required function (mem service) in FX.h?
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX_fcn.cpp:13]: AFAIK, no. I'll leave PS to you entirely, even though I have modified version in my fork..

File: wled00/data/index.js [hunk 12]
```diff
@@ -35,9 +35,10 @@ var cfg = {
 // [year, month (0 -> January, 11 -> December), day, duration in days, image url]
 var hol = [
 	[0, 11, 24, 4, "https://aircoookie.github.io/xmas.png"],		// christmas
-	[0, 2, 17, 1, "https://images.alphacoders.com/491/491123.jpg"], // st. Patrick's day
-	[2025, 3, 20, 2, "https://aircoookie.github.io/easter.png"],	// easter 2025
-	[2024, 2, 31, 2, "https://aircoookie.github.io/easter.png"],	// easter 2024
+	[0, 2, 17, 1, "https://images.alphacoders.com/491/491123.jpg"],	// st. Patrick's day
+	[2026, 3, 5, 2, "https://aircoookie.github.io/easter.png"],		// easter 2026
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/data/index.js:39]: Tabs vs spaces. 👍

File: wled00/colors.cpp [hunk 13]
```diff
@@ -208,14 +208,14 @@ CRGBPalette16 generateHarmonicRandomPalette(const CRGBPalette16 &basepalette)
...
-  // apply saturation & gamma correction
+  // apply saturation
   CRGB RGBpalettecolors[4];
   for (int i = 0; i < 4; i++) {
     if (makepastelpalette && palettecolors[i].saturation > 180) {
       palettecolors[i].saturation -= 160; //desaturate all four colors
     }
     RGBpalettecolors[i] = (CRGB)palettecolors[i]; //convert to RGB
-    RGBpalettecolors[i] = gamma32(((uint32_t)RGBpalettecolors[i]) & 0x00FFFFFFU); //strip alpha from CRGB
+    RGBpalettecolors[i] = ((uint32_t)RGBpalettecolors[i]) & 0x00FFFFFFU; //strip alpha from CRGB
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/colors.cpp:218]: I was just removing gamma32() using search and replace. I did not delve deeply what the above code actually does.

File: wled00/util.cpp [hunk 14]
```diff
@@ -619,6 +619,68 @@ int32_t hw_random(int32_t lowerlimit, int32_t upperlimit) {
   return hw_random(diff) + lowerlimit;
 }
 
+#ifndef ESP8266
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/util.cpp:622]: It is not. You can compile with the ability to use PSRAM but if it is actually available is a runtime detection.

File: wled00/FX.h [hunk 15]
```diff
@@ -387,239 +412,250 @@ typedef enum mapping1D2D {
...
-      _t(nullptr)
-    {
-      #ifdef WLED_DEBUG
-      //Serial.printf("-- Creating segment: %p\n", this);
-      #endif
+    inline uint32_t *getPixels() const                              { return pixels; }
+    inline void     setPixelColorRaw(unsigned i, uint32_t c) const  { pixels[i] = c; }
+    inline uint32_t getPixelColorRaw(unsigned i) const              { return pixels[i]; };
+  #ifndef WLED_DISABLE_2D
+    inline void     setPixelColorXYRaw(unsigned x, unsigned y, uint32_t c) const  { auto XY = [](unsigned X, unsigned Y){ return X + Y*Segment::vWidth(); }; pixels[XY(x,y)] = c; }
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX.h:541]: Lambdas always compile as inline.

File: wled00/FX.h [hunk 16]
```diff
@@ -644,66 +678,44 @@ typedef struct Segment {
...
 
     // 1D strip
-    [[gnu::hot]] uint16_t virtualLength() const;
-    [[gnu::hot]] void setPixelColor(int i, uint32_t c) const; // set relative pixel within segment with color
+    uint16_t virtualLength() const;
+    [[gnu::hot]] void setPixelColor(int n, uint32_t c) const; // set relative pixel within segment with color
     inline void setPixelColor(unsigned n, uint32_t c) const                    { setPixelColor(int(n), c); }
     inline void setPixelColor(int n, byte r, byte g, byte b, byte w = 0) const { setPixelColor(n, RGBW32(r,g,b,w)); }
     inline void setPixelColor(int n, CRGB c) const                             { setPixelColor(n, RGBW32(c.r,c.g,c.b,0)); }
+    void setRawPixelColor(int i, uint32_t col) const                           { if (i >= 0 && i < length()) setPixelColorRaw(i,col); }
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX.h:691]: Some effects expect indexes to be negative and rely on sPC to filter out these. If you want speed in PS use sPCRaw().

I do not approve implicit conversion of negative int into unsigned. Speed or no speed.

File: wled00/FX_fcn.cpp (hunk 4 above)
Comment (blazoncek) [reply wled00/FX_fcn.cpp:885]: Intentionally no. See answer above.

File: wled00/FX_fcn.cpp (hunk 8 above)
Comment (blazoncek) [reply wled00/FX_fcn.cpp:1074]: IDK I did no comparison. Rainbow mode is visibly different but changing this may affect final output annoying (some) users.

File: wled00/FX.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX.cpp:6168]: yes, but without overlay, it was adding a trail (or motion blur). I don't think adding motion blur should always be there, on fast moving text it will be "unclean"

File: wled00/FX_fcn.cpp (hunk 6 above)
Comment (blazoncek) [reply wled00/FX_fcn.cpp:1037]: An oversight, possibly, as not everything from upstream is included in my fork.

File: wled00/FX.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX.h:691]: you can also use an explicit conversion. I do not like the extra operation in such a key function just for cosmetic reasons.

File: wled00/FX.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX.h:541]: yes sure, its just convoluted and does not improve readability IMHO.

File: wled00/util.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/util.cpp:622]: AFAIK there aren't any C3 boards with PSRAM. Would it be better to add C3 to the #ifndef?

File: wled00/FX.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX.cpp:5057]: pal 35 IS a predefined palette (fire). with loadPalette() removed, how can an FX use a certain fixed palette i.e. what is the new way to do this?

[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX.cpp:5057]: There is `SEGMENT.color_from_palette()` or FastLED's `ColorFromPalette()` (or your implementation of it) with proper ID.

File: wled00/FX_fcn.cpp [hunk 17]
```diff
@@ -137,48 +126,60 @@ Segment& Segment::operator= (const Segment &orig) {
...
   if (data && _dataLen >= len) {          // already allocated enough (reduce fragmentation)
-    if (call == 0) memset(data, 0, len);  // erase buffer if called during effect initialisation
+    if (call == 0) {
+      //DEBUG_PRINTF_P(PSTR("--   Clearing data (%d): %p\n"), len, this);
+      memset(data, 0, len);  // erase buffer if called during effect initialisation
+    }
     return true;
   }
-  //DEBUG_PRINTF_P(PSTR("--   Allocating data (%d): %p\n", len, this);
-  deallocateData(); // if the old buffer was smaller release it first
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/FX_fcn.cpp:160]: Consider heap fragmentation and the tradeoff of keeping data allocated.

File: wled00/fcn_declare.h [hunk 18]
```diff
@@ -544,6 +543,27 @@ inline uint8_t hw_random8() { return HW_RND_REGISTER; };
 inline uint8_t hw_random8(uint32_t upperlimit) { return (hw_random8() * upperlimit) >> 8; }; // input range 0-255
 inline uint8_t hw_random8(uint32_t lowerlimit, uint32_t upperlimit) { uint32_t range = upperlimit - lowerlimit; return lowerlimit + hw_random8(range); }; // input range 0-255
 
+// PSRAM allocation wrappers
+#ifndef ESP8266
+void *w_malloc(size_t);           // prefer PSRAM over DRAM
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/fcn_declare.h:548]: It started with w (as in WLED's malloc). Later I added d variant.

File: wled00/FX.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/FX.cpp:6168]: please disregard, I completely ignored that custom1 slider is purely for this purpose and its default is 0.

**Review by DedeHai — CHANGES_REQUESTED** [review]

**Comment by DedeHai** [comment]
I played around with it a bit:
- FPS calculation needs reverting (see inline comments)
- the new gamma correction really does not play well with the PS: 
   - if gamma correction is disabled, palettes are ugly
   - if it is enabled, particle movement is very choppy (overall brightness is screwed up)
   - I don't see a way to fix this: if gamma is applied at the end (as would be the correct way) the brightness of a palette pixel changes: gamma should be applied to full brightness palette color.

File: wled00/const.h [hunk 19]
```diff
@@ -151,6 +137,8 @@
   #endif
 #endif
 
+#define WLED_MAX_PANELS 18                      // must not be more than 32
+
```
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/const.h:141]: This is unoverridable macro so no need for guard.

**Comment by DedeHai** [comment]
any ideas how to tackle the gamma correction issue?

**Comment by blazoncek** [comment]
For correctness of display gamma correction should (or must) be applied as a last step in image processing. Otherwise you will get non-linear pixel manipulation which will produce incorrect results.

Originally, WLED applied gamma immediately to color only, palettes had gamma (variable!!!) pre-applied. This produced, as you noted, some weird effects when brightness was not 255 (either as segment opacity or global brightness) and effect used previous pixel values as a base for next frame or neighbouring pixels.

@Makuna implemented `NeoPixelBusLg` (as an evolution of `NeoPixelBrightnessBus` that WLED used originally) that applies gamma while outputting to pixels (albeit currently still modifying pixel buffer, but AFAIK that will change in one of the future versions). As WLED had palettes with pre-applied gamma that feature was unusable and we only used brightness/lightness component of it otherwise we'd be using it since the switch.

Moving gamma correction to the last step in image creation is correct way of doing things and I don't think we need to "address" it.

**Comment by DedeHai** [comment]
While applying gamma as a last step is correct, it does alter the overall look of many FX. 

> the actual problem is this: the PS spreads the brightness of a particle over 4 pixels when rendering. this MUST be linear for the total brightness to be constant. If gamma is applied afterwards, this becomes an exponential function and the particle brightness changes depending on sub-pixel position, making the movements look choppy.

> this problem does not only apply to the PS but to all FX that use scaled palette colors, i.e. brightness < 255 in colorFromPalette() call: applying gamma at the end (while correct in theory) changes the look of many FX: look at pacifica for example. So unless we can find a good solution, the gamma correction needs to be reverted.

this needs a solution

**Comment by blazoncek** [comment]
If PS has problems and assumes gamma corrected colors, then it needs update. You said "MUST be linear", with gamma pre-applied any operation on pixels is not linear.
Pacifica (and a few other effects) use custom palette(s) for color generation, those need to be adapted to not be gamma adjusted (or inverse gamma applied to the values), as I did for all other 50+ palettes. *Different look* of some effects was already something that "global buffer" introduced as color restoration was no longer needed (this produced different look on ESP8266 and ESP32).

However, if you insist on gamma to be applied before any other processing, it can still be done in appropriate function calls.
I.e. `Segment::getCurrentColor()`, `ColorFromPaletteWLED()` and `Segment::color_from_palette()` with minimal impact.

A side note: Gamma use was not entirely consistent in WLED, take for example `Segment::color_wheel()` or look at the old code. This PR *enforces* consistent gamma use across every aspect (as it is not used anywhere but sending pixel data to display).

**Comment by DedeHai** [comment]
> If PS has problems and assumes gamma corrected colors, then it needs update. You said "MUST be linear", with gamma pre-applied any operation on pixels is not linear.

brightness is still linear, which is the point.
also: to be entirely correct, gamma must be applied as the last step AFTER applying any brightness scaling. If I understand the code right, brightness is applied by NPB (and ABL) which is done after gamma correction right? So if brightness is not at 100%, correction is still not entirely right. 
I played with some of the FX and the looks is noticeably different. There is no option for users to get the old looks back so I expect complaints.

As for the PS: to make it work with the new gamma correction, this is what is needed:
- calculate an inverse-gamma table, using 256 bytes of RAM
- apply inverse gamma pre-correction to every pixel of every particle: will give a slight speed penalty but should not be huge

**Comment by DedeHai** [comment]
I ran some tests and while scaling globally after applying gamma does not preserve _perceived_ colors, it actually makes dark colors brighter, which for WLED purposes is probably better: pixels do not fade to black as fast in darker regions at low brightness settings. So I agree that the implementation is correct.

Just to explain what I mean by "to be entirely correct, gamma must be applied as the last step AFTER applying any brightness scaling." let me run an example on what I mean using gamma=2.0 :

assuming a color value of 200: gamma corrected output value for this color is 155
scaling it to 50% brightness _before_ gamma correction (i.e. gamma(100)): output is 38, meaning a LED value of 38 is perceived as 50% of 155
scaling it to 50% brightness _after_gamma correction: output is 155/2 = 77 corresponding to a perceived brightness of 141 (i.e. gamma(141) = 77) or 90% of the original brightness. 

adding gamma for brightness (if set) after gamma correction does indeed correct for this, so you are right that it's up to the user which one they prefer.

As for the PS: I did add an inverse gamma-table and it fixes the brightness distribution at the cost of 2% FPS drop. Given that direct rendering to the buffer will more than make up for that its not an issue.

Should I push the updated PS code to this PR?

File: wled00/image_loader.cpp [hunk 20]
```diff
@@ -78,7 +78,7 @@ void drawPixelCallback(int16_t x, int16_t y, uint8_t red, uint8_t green, uint8_t
 byte renderImageToSegment(Segment &seg) {
   if (!seg.name) return IMAGE_ERROR_NO_NAME;
   // disable during effect transition, causes flickering, multiple allocations and depending on image, part of old FX remaining
-  if (seg.mode != seg.currentMode()) return IMAGE_ERROR_WAITING;
+  //if (seg.mode != seg.currentMode()) return IMAGE_ERROR_WAITING;
```
Comment (DedeHai) [inline wled00/image_loader.cpp:81]: not sure what it is but gif player is broken, there is no output, its all black.
also: in line 61 above: should the gamma correction be removed?

**Comment by blazoncek** [comment]
@DedeHai why would you add *inverse gamma correction*? Gamma is not applied until everything is rendered, including overlays and live (i.e DDP/ArtNet/E1.31 data).
BTW, wait with other changes until this is merged IMO. I know it may break nightly builds but that may be expected.

As for GIF image effect, I do not know it, never run it, so would need @Aircoookie to help with it (possibly after this is merged or if @Aircoookie adds to this PR).

**Comment by DedeHai** [comment]
> why would you add _inverse gamma correction_?

like I explained above. I need linear total brightness distribution, adding gamma does not allow that unless its inverted.

**Comment by blazoncek** [comment]
> like I explained above.

I read it at least twice or perhaps even 3 or more times but I still fail to understand why *inverse* gamma. I would understand if you'd use *inverse* gamma in the past (where gamma was applied to palettes and colors immediately upon selection) when colors were gamma adjusted at the start of rendering process.

I am getting the impression that we are talking about two different things. I guess you want to keep the looks of existing implementation, not necessarily directly linked to the place of gamma application, but instead based on assumption that current behaviour is the "correct" one (or the one that should be mimicked).

**Comment by DedeHai** [comment]
just ran a side-by-side comparison of main and this PR: found the best match in looks is with gamma = 1.8 or there abouts, maybe 1.9. If you can: run some side by side comparison for yourself but I think the default gamma should be changed to 1.8 or 1.9

**Comment by blazoncek** [comment]
> apply gamma to that and it is no longer constant but depends on distribution.

I think I'm starting to understand what your problem is. 😀 It is not gamma but your perception.

I'll try to explain. Gamma is used to adjust perceived brightness to be linear (i.e. brightness at low levels is perceived much brighter than its value would suggest on LEDs) so that you can work with linear math when modifying pixel buffer (LEDs will receive gamma adjusted values from pixel buffer). As such PS need not take into account any gamma. Yet, you will argue that applying gamma (as a last step) will make PS effect look *ugly* (or incorrect) while in reality *it is what is intended* by using gamma. Remember, this is only effective while outputting to LEDs, so all pixel data is unaffected by gamma and needs no adjustment (including subpixel calculations).

Perhaps, due to the change when and where gamma is applied, it might be better to reverse recommendation and instead of recommending gamma for color we should recommend gamma for brightness and reduce the default value of 2.8 to something lower i.e 2.2 (I'm using this value on my installs). Or, perhaps, disable gamma altogether on new installs.

**Comment by DedeHai** [comment]
I do understand that gamma is used to convert _linear_ color brightness to _perceived_ brightness. For the particle rendering however, the sum of brightnesses in the sense of "number of photons" must be constant as a particle does not change in total light output, independent of the brightness distribution on those four pixels (or two in 1D). The "number of photons" is (more or less) linear to the LED PWM value sent to the LEDs.
Example for two pixels:
(128,128) -> sum is 256 -> apply gamma: (56,56) -> sum is 112 
lets move the particle a little and recalculate:
(100,156) -> sum is 256 -> apply gamma: (33, 87) -> sum is 120
so the particle appears brighter as the total amount of light increases as it moves away from the center between two pixels. this gets more extreme the closer it comes to the "edge" before jumping to the next pixel.
so even though the perceived brightness is mapped correctly, it does not keep the total amount of light constant if that is distributed linearly. 
now my assumption may not be 100% correct on how the eye perceives a "set of four pixels" in total brightness. But looking at PS animations, not doing any gamma inversion results in noticeable brightness fluctuations as a particle moves.
take an edge case as a comparison:
(20,236) -> sum is 256 -> apply gamma: (1,215) -> sum is 216
so in this case the total light output is almost double than when evenly distributed, the perceived brightness is of course not doubled but it is noticeably brighter.

with the palettes having gamma removed, applying gamma correction is not optional really: take a look at fire palette with gamma disabled, it is very much not like fire. Applying gamma only to brightness will not fix that.

**Comment by blazoncek** [comment]
I give up.

**Comment by DedeHai** [comment]
just look at a PS effect that has single particles, like ballpit at lowest speed and size=1 and you will see what I mean.

is it possible to move the new gamma correction to a seperate PR? I feel like that needs more testing by more people.

**Comment by DedeHai** [comment]
I ran a few more palette tests: 
- seaviewsensing palettes were not converted correctly: if applying gamma of 2.2, the palettes need to have additional initial gamma correction: 1.182 for red, 1.0 for green, 1.1364 for blue to match the palettes to match the original 2.6, 2.2, 2.5 correction
- other palettes (like orangery) need to have their gamma reversed to match color after applying gamma 2.2

the difference in orangery for example is quite big between old and new correction.

**Review by DedeHai — APPROVED** [review]
approved. As discussed, some cleanup will be needed after merge
- all palettes from seaviewsensing need some initial correction to match original looks (see comment above)
- PS requires gamma inversion for brightness distribution calculation (fix ready)
- Some FX need tuning to not change too much. I have some work in progress on Pacifica, Aurora is another candidate, there are probably a few others

**Comment by blazoncek** [comment]
Does anyone else have any considerations? Should I merge or wait still?

**Comment by willmmiles** [comment]
I recommend merging and adding an issue entry (milestone 0.16.0, keep) to track any outstanding gamma adjustments.
